app/controllers/admin/api/service_contracts_controller.rb

```diff
@@ -6,29 +6,91 @@ class Admin::Api::ServiceContractsController < Admin::Api::ServiceBaseController
 
   before_action :deny_on_premises_for_master
   before_action :authorize_service_plans!
+  before_action :find_account
+  before_action :find_service_contract, only: [:show, :update, :destroy]
+
+
+  # Service Subscription Create
+  # POST /admin/api/accounts/:account_id/service_contracts.xml
+  def create
+    plan = service_plan
+
+    unless valid_plan_and_account?(plan)
+      render_error('Invalid service plan or account id', status: :unprocessable_entity)
+      return
+    end
+
+    create_service_contract(plan)
+  end
 
   # Service Subscription List
   # GET /admin/api/accounts/{account_id}/service_contracts.xml
   def index
-    respond_with account.bought_service_contracts
+    respond_with @account.bought_service_contracts
   end
 
   # Service Subscription Delete
   # DELETE /admin/api/accounts/{account_id}/service_contracts/{id}.xml
   def destroy
-    service_subscription = ServiceSubscriptionService.new(account)
+    service_subscription = ServiceSubscriptionService.new(@account)
 
-    respond_with(service_subscription.unsubscribe(service_contract))
+    respond_with(service_subscription.unsubscribe(@service_contract))
+  end
+
+  # Service Subscription Update
+  # PUT /admin/api/accounts/{account_id}/service_contracts/{id}.xml
+  def update
+    service = @service_contract.issuer
+    new_plan = service.service_plans.find(service_contract_plan_id)
+    @service_contract.change_plan!(new_plan)
+    respond_with(@service_contract)
+  end
+
+  # Service Subscription Show
+  # GET /admin/api/accounts/:account_id/service_contracts/:id.xml
+  def show
+    respond_with @service_contract
   end
 
   protected
 
-  def account
-    @account ||= current_account.buyers.find params.require(:account_id)
+  def find_account
+    @account ||= current_account.buyers.find_by(id: params.require(:account_id))
+
+    unless @account
+      render_error('Buyer not found with this account ID', status: :not_found)
+    end
```

Comment:
```suggestion
  def account
    @account ||= current_account.buyers.find params.require(:account_id)
  end
```

I think this change is not needed, the original code was working just fine.
So, `before_action :find_account` is not needed either, and `@account` can be changed to `account`.

`#find` method throws a `ActiveRecord::RecordNotFound`, which is handled [here](https://github.com/3scale/porta/blob/e50efabc23679892d43115a9d2238e864ead7d93/app/controllers/concerns/error_handling.rb#L8), that makes the API request to return 404.

I think we don't need to include the error message (and moreover we should avoid the term "buyer" whenever possible), because all other API endpoints just return 404 with no error message.

Reply:
Sounds good but I added so whoever use API can clearly know okay this input is wrong